File: src/Z.Collections/System.Collections.Generic.IDictionary[string, object]/IDictionary[string, object].ToExpando.cs
// Description: C# Extension Methods | Enhance the .NET Framework and .NET Core with over 1000 extension methods.
// Website & Documentation: https://csharp-extension.com/
// Issues: https://github.com/zzzprojects/Z.ExtensionMethods/issues
// License (MIT): https://github.com/zzzprojects/Z.ExtensionMethods/blob/master/LICENSE
// More projects: https://zzzprojects.com/
// Copyright © ZZZ Projects Inc. All rights reserved.
using System.Collections.Generic;
using System.Dynamic;

public static partial class Extensions
{
    /// <summary>
    ///     An IDictionary&lt;string,object&gt; extension method that converts the @this to an expando.
    /// </summary>
    /// <param name="this">The @this to act on.</param>
    /// <returns>@this as an ExpandoObject.</returns>
    public static ExpandoObject ToExpando(this IDictionary<string, object> @this)
    {
        var expando = new ExpandoObject();
        var expandoDict = (IDictionary<string, object>) expando;

        foreach (var item in @this)
        {
            if (item.Value is IDictionary<string, object>)
            {
                var d = (IDictionary<string, object>) item.Value;
                expandoDict.Add(item.Key, d.ToExpando());
            }
            else
            {
                expandoDict.Add(item);
            }
        }

        return expando;
    }
}

File: src (VB.NET)/Z.ExtensionMethods.VB/Z.Collections/System.Collections.Generic.IDictionary[string, object]/IDictionary[string, object].ToExpando.vb

' Copyright (c) 2015 ZZZ Projects. All rights reserved
' Licensed under MIT License (MIT) (https://github.com/zzzprojects/Z.ExtensionMethods)
' Website: http://www.zzzprojects.com/
' Feedback / Feature Requests / Issues : http://zzzprojects.uservoice.com/forums/283927
' All ZZZ Projects products: Entity Framework Extensions / Bulk Operations / Extension Methods /Icon Library

Imports System.Collections.Generic
Imports System.Dynamic

Public Module Extensions_19

	''' <summary>
	'''     An IDictionary&lt;string,object&gt; extension method that converts the @this to an expando.
	''' </summary>
	''' <param name="this">The @this to act on.</param>
	''' <returns>@this as an ExpandoObject.</returns>
	<System.Runtime.CompilerServices.Extension> _
	Public Function ToExpando(this As IDictionary(Of String, Object)) As ExpandoObject
		Dim expando = New ExpandoObject()
		Dim expandoDict = DirectCast(expando, IDictionary(Of String, Object))

        For Each item As Object In this
            If TypeOf item.Value Is IDictionary(Of String, Object) Then
                Dim d = DirectCast(item.Value, IDictionary(Of String, Object))
                expandoDict.Add(item.Key, d.ToExpando())
            Else
                expandoDict.Add(item)
            End If
        Next

		Return expando
	End Function
End Module


